data_structures/src/mainnet_validations.rs

```diff
@@ -16,6 +18,145 @@ pub const FIRST_HARD_FORK: Epoch = 192000;
 /// 28 April 2021 @ 9:00:00 UTC
 pub const SECOND_HARD_FORK: Epoch = 376320;
 
+/// TAPI Engine
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct TapiEngine {
+    /// bit votes counter by bits
+    pub bit_tapi_counter: BitTapiCounter,
+    /// wip activation
+    pub wip_activation: HashMap<String, Epoch>,
+}
+
+impl TapiEngine {
+    pub fn update_bit_counter(&mut self, v: u32, epoch: Epoch, avoid_wip_list: &HashSet<String>) {
+        // In case of empty epochs, they would be considered as blocks with tapi version to 0
```

Comment:
Can you add an assert here so we can catch errors when testing?

`assert!(epoch > self.bit_tapi_counter.last_epoch);`

Reply:
Ah, but maybe this conflicts with the function `update_new_wip_votes` because there we want to update old epochs?